File: src/main/java/space/accident/api/interfaces/tileentity/ICoverable.java
package space.accident.api.interfaces.tileentity;

import net.minecraft.entity.player.EntityPlayerMP;
import net.minecraft.item.ItemStack;
import space.accident.api.interfaces.tileentity.energy.IEnergyTile;
import space.accident.api.interfaces.tileentity.redstone.IRedStoneTileEntity;
import space.accident.api.util.SA_CoverBehaviorBase;
import space.accident.api.util.ISerializableObject;

public interface ICoverable extends IRedStoneTileEntity, IHasInventory, IEnergyTile {
	boolean canPlaceCoverIDAtSide(int side, int id);
	
	boolean canPlaceCoverItemAtSide(int side, ItemStack aCover);
	
	boolean dropCover(int side, int aDroppedSide, boolean aForced);
	
	void setCoverDataAtSide(int side, ISerializableObject aData);
	
	void setCoverIdAndDataAtSide(int side, int id, ISerializableObject aData);
	void setCoverIDAtSide(int side, int id);
	boolean setCoverIDAtSideNoUpdate(int side, int id);
	void setCoverItemAtSide(int side, ItemStack aCover);
	
	ISerializableObject getComplexCoverDataAtSide(int side);
	
	int getCoverIDAtSide(int side);
	
	ItemStack getCoverItemAtSide(int side);
	
	SA_CoverBehaviorBase<?> getCoverBehaviorAtSideNew(int side);
	
	/**
	 * For use by the regular MetaTileEntities. Returns the Cover Manipulated input RedStone.
	 * Don't use this if you are a Cover Behavior. Only for MetaTileEntities.
	 */
	int getInternalInputRedStoneSignal(int side);
	
	/**
	 * For use by the regular MetaTileEntities. This makes it not conflict with Cover based RedStone Signals.
	 * Don't use this if you are a Cover Behavior. Only for MetaTileEntities.
	 */
	void setInternalOutputRedStoneSignal(int side, int aStrength);
	
	/**
	 * Causes a general Cover Texture update.
	 * Sends 6 Integers to Client + causes @issueTextureUpdate()
	 */
	void issueCoverUpdate(int side);
	
	/**
	 * Receiving a packet with cover data.
	 */
	void receiveCoverData(int coverSide, int coverID, int coverData);
	
	/**
	 * Receiving a packet with cover data.
	 * @param player the player who made the change
	 */
	default void receiveCoverData(int coverSide, int coverId, ISerializableObject coverData, EntityPlayerMP player) {
		if (coverData instanceof ISerializableObject.LegacyCoverData)
			receiveCoverData(coverSide, coverId, ((ISerializableObject.LegacyCoverData) coverData).get());
	}
}

File: src/main/kotlin/space/accident/api/interfaces/te/ICoverable.kt
package space.accident.api.interfaces.te

import net.minecraft.entity.player.EntityPlayerMP
import net.minecraft.item.ItemStack
import space.accident.api.util.GT_CoverBehaviorBase
import space.accident.api.util.ISerializableObject

interface ICoverable : IRedstoneTileEntity, IHasInventory, IBasicEnergyContainer {

    fun canPlaceCoverIDAtSide(aSide: Byte, aID: Int): Boolean
    fun canPlaceCoverItemAtSide(aSide: Byte, aCover: ItemStack?): Boolean
    fun dropCover(aSide: Byte, aDroppedSide: Byte, aForced: Boolean): Boolean

    fun setCoverDataAtSide(aSide: Byte, aData: ISerializableObject)
    fun setCoverIdAndDataAtSide(aSide: Byte, aId: Int, aData: ISerializableObject?)
    fun setCoverIDAtSide(aSide: Byte, aID: Int)
    fun setCoverIDAtSideNoUpdate(aSide: Byte, aID: Int): Boolean
    fun setCoverItemAtSide(aSide: Byte, aCover: ItemStack?)

    fun getComplexCoverDataAtSide(aSide: Byte): ISerializableObject

    fun getCoverIDAtSide(aSide: Byte): Int
    fun getCoverItemAtSide(aSide: Byte): ItemStack?


    fun <CoverBase : ISerializableObject> getCoverBehaviorAtSideNew(aSide: Byte): GT_CoverBehaviorBase<CoverBase>

    /**
     * For use by the regular MetaTileEntities. Returns the Cover Manipulated input Redstone.
     * Don't use this if you are a Cover Behavior. Only for MetaTileEntities.
     */
    fun getInternalInputRedstoneSignal(aSide: Byte): Byte

    /**
     * For use by the regular MetaTileEntities. This makes it not conflict with Cover based Redstone Signals.
     * Don't use this if you are a Cover Behavior. Only for MetaTileEntities.
     */
    fun setInternalOutputRedstoneSignal(aSide: Byte, aStrength: Byte)

    /**
     * Causes a general Cover Texture update.
     * Sends 6 Integers to Client + causes @issueTextureUpdate()
     */
    fun issueCoverUpdate(aSide: Byte)

    /**
     * Receiving a packet with cover data.
     */
    fun receiveCoverData(coverSide: Byte, coverID: Int, coverData: Int)

    /**
     * Receiving a packet with cover data.
     * @param aPlayer the player who made the change
     */
    fun receiveCoverData(aCoverSide: Byte, aCoverID: Int, aCoverData: ISerializableObject, aPlayer: EntityPlayerMP?) {
        if (aCoverData is ISerializableObject.LegacyCoverData) {
            receiveCoverData(aCoverSide, aCoverID, aCoverData.get())
        }
    }
}
